GVFS/GVFS.Mount/InProcessMount.cs

```diff
@@ -299,11 +304,25 @@ private void HandleDehydrateFolders(NamedPipeMessages.Message message, NamedPipe
                     {
                         response.FailedFolders.Add($"{folder}\0{errorMessage}");
                     }
+
+                    resetFolderPaths.Append($"\"{folder.Replace(Path.DirectorySeparatorChar, GVFSConstants.GitPathSeparator)}\" ");
                 }
 
-                // Since placeholders and modified paths could have changed with the dehydrate, the index needs to be rebuilt
+                // Since modified paths could have changed with the dehydrate, the paths that were dehydrated need to be reset in the index
+                string resetPaths = resetFolderPaths.ToString();
                 GitProcess gitProcess = new GitProcess(this.enlistment);
-                gitProcess.ForceCheckout(GVFSConstants.DotGit.HeadName);
+                GitProcess.Result refreshIndexResult = gitProcess.Reset(GVFSConstants.DotGit.HeadName, resetPaths);

```

Comment:
What do we do if this fails?  What state will things be in?  Will the user be able to recover and is there some message we can surface to the user?  I realize we weren't even checking the result before but we are now allowing this to run when the user doesn't have a clean status.

Reply:
All great questions!

> What state will things be in?

Assuming that the reset completely fails (leaving things in the same state that they were in before) users will be unable to see some or all of the contents of the folder(s) that were just dehydrated.  The folder (or some of its contents) will be missing because:

- VFS4G only projects files whose skip-worktree bit is set
- The failure of `git reset` means that the entries in the index were not updated to have their skip worktree bit set.

Additionally, `git status` would not report these files as deleted because they are not in ModifiedPaths.dat.

> Will the user be able to recover and is there some message we can surface to the user?

It is possible to recover from this situation, but with the current code knowing how to recover (or even that recovering is required) would be difficult for most users.  We could report back that this portion of the `dehydrate` failed, but I think we'll also have to decide what step(s) to ask the user to run.

> What do we do if this fails?

Related to the question of what we surface to the user, I think we have a few options:

- If we're willing to take a perf hit, we could dehydrate/reset each folder one at a time.  If the dehydrate fails we could add the failed folder to ModifiedPaths.dat and ask the user to try again.

- If we want to avoid lots of calls to `git reset` we could be pessimistic and all of the folders that were dehydrated to ModifiedPaths.dat and ask the user to run the command again.

- We could tell the user that VFS4G failed to update the index, and ask them commit any changes they may have and run `git checkout HEAD`.  I am not sure if this advice will always be 100% good though, as if `dehydrate` is called from a post-command hook a user could find themselves in this situation while another git command (e.g. merge, rebase, etc.) is still in progress.

I am leaning toward option 2 (possibly as a follow up PR), but I'd like to hear if you can think of any alternatives.  

CC: @derrickstolee for his thoughts on how we can avoid straying too far from native git sparse-checkout behavior in similar situations.

Reply:
Filed #1606 for this work.

Reply:
Yeah option 2 does seem like the best. Anything we can do to help the user know what to do but also not make to big an impact on performance would be best.